qiskit/qobj/_qobj.py

```diff
@@ -0,0 +1,209 @@
+        id (str): QObj identifier.
+        config (QObjConfig): config settings for the QObj.
+        circuits (list[QObjExperiment]): list of experiments.
+        type (str): experiment type (QASM/PULSE).
+        _version (str): QObj version.
+    """
+
+    REQUIRED_ARGS = ['id', 'config', 'circuits']
+
+    def __init__(self, id, config, circuits, **kwargs):
+        # pylint: disable=redefined-builtin,invalid-name
+        self.id = id
+        self.config = config
+        self.circuits = circuits
+
+        self.type = QObjType.QASM.value
+        self._version = QOBJ_VERSION
+
+        super().__init__(**kwargs)
+
+
+class QObjConfig(QObjItem):
+    """Configuration for a QObj.
+
+    Attributes:
+        max_credits (int): number of credits.
+        shots (int): number of shots.
+        backend_name (str): name of the backend.
+    """
+    REQUIRED_ARGS = ['max_credits', 'shots', 'backend_name']
+
+    def __init__(self, max_credits, shots, backend_name, **kwargs):
+        self.max_credits = max_credits
+        self.shots = shots
+        self.backend_name = backend_name
+
+        super().__init__(**kwargs)
+
+
+class QObjExperiment(QObjItem):
+    """Quantum experiment represented inside a QObj.
+
+    Attributes:
+        name (str): name of the experiment.
+        config (QObjExperimentConfig): config settings for the experiment.
+        compiled_circuit (QObjCompiledCircuit): list of instructions
+        compiled_circuit_qasm (str)
+    """
+    REQUIRED_ARGS = ['name', 'config', 'compiled_circuit',
+                     'compiled_circuit_qasm']
+
+    def __init__(self, name, config, compiled_circuit, compiled_circuit_qasm,
+                 **kwargs):
+        self.name = name
+        self.config = config
+        self.compiled_circuit = compiled_circuit
+        self.compiled_circuit_qasm = compiled_circuit_qasm
+
+        super().__init__(**kwargs)
+
+
+class QObjInstruction(QObjItem):
+    """Quantum Instruction.
+
+    Attributes:
+        name(str): name of the gate.
+        qubits(list): list of qubits to apply to the gate.
+    """
+    REQUIRED_ARGS = ['name', 'qubits']
+
+    def __init__(self, name, qubits, **kwargs):
+        self.name = name
+        self.qubits = qubits
+
+        super().__init__(**kwargs)
+
+
+# TODO: Remove when new schema is in place?
+class QObjExperimentConfig(QObjItem):
+    """Configuration for a experiment.
+
+    Attributes:
+        seed (int): seed.
+        basis_gates (str): basis gates
+        coupling_map (list): coupling map
+        layout (list): layout
+    """
+    REQUIRED_ARGS = ['seed', 'basis_gates', 'coupling_map', 'layout']
+
+    def __init__(self, seed, basis_gates, coupling_map, layout, **kwargs):
+        self.seed = seed
+        self.basis_gates = basis_gates
+        self.coupling_map = coupling_map
+        self.layout = layout
+
+        super().__init__(**kwargs)
+
+
+# TODO: Remove when new schema is in place?
+class QObjCompiledCircuit(QObjItem):
```

Comment:
This class should be called `QobjCompiledExperiment`. The idea with qobj is that it is a datastructure describing what the backend should do. And this could be either in the language of circuits/gates, or in the language of pulses. Experiment is a generic word that captures both.

I know that in our current setup we have things like `compiled_circuit` and `compiled_circuit_qasm`, and these should stay until full migration (as you have done in your `QobjExperiment` class). But we can at least define the classes now, as we would like them in the future.

Reply:
Also, I think I don't agree with your `TODO` here. This class should stay, as it is the main class describing the payload of the experiment to be executed.